File: 0097_Interleaving String/InterleavingString_DP.cs
public class Solution {
    public bool IsInterleave(string s1, string s2, string s3) {
        var l1 = s1.Length;
        var l2 = s2.Length;
        var l3 = s3.Length;
        if(l1 + l2 != l3) return false;
        // dp[i,j]: whehter s3[0:i+j] is a interleaving of s1[0:i] and s2[0:j]

        bool[,] dp = new bool[l1+1, l2+1];
        dp[0,0] = true;
        for(var i=0;i<=l1;i++)
        {
            for(var j=0;j<=l2;j++)
            {
                if(i>0 && s1[i-1] == s3[i+j-1])
                    dp[i,j] = dp[i-1,j];
                
                if(!dp[i,j])
                {
                    if(j>0 && s2[j-1] == s3[i+j-1])
                        dp[i,j] = dp[i,j-1];
                }
            }
        }
        
        return dp[l1,l2];
    }
}

File: 0097_Interleaving String/InterleavingString_DP.java
public class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        int m = s1.length();
        int n = s2.length();
        int l = s3.length();
        if(m + n != l) return false;
        
        boolean[][] dp = new boolean[m+1][n+1];
        
        for (int i = 0; i <= m; i++) {
            for (int j = 0; j <= n; j++) {
                // both s1 and s2 are empty string
                if (i == 0 && j == 0) {
                    dp[i][j] = true;
                // when s1 is empty string, only compare s2 with s3
                } else if (i == 0) {
                    dp[i][j] = dp[i][j - 1] && s2.charAt(j - 1) == s3.charAt(i + j - 1);
                // when s2 is empty string, only compare s1 with s3
                } else if (j == 0) {
                    dp[i][j] = dp[i - 1][j] && s1.charAt(i - 1) == s3.charAt(i + j - 1);
                } else {
                    dp[i][j] = (dp[i - 1][j] && s1.charAt(i - 1) == s3.charAt(i + j - 1)) || (dp[i][j - 1] && s2.charAt(j - 1) == s3.charAt(i + j - 1));
                }
            }
        }
        return dp[m][n];
    }
}